pkg/certwatcher/watcher.go

```diff
@@ -0,0 +1,79 @@
+package certwatcher
+
+import (
+	"context"
+	"crypto/tls"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"go.uber.org/zap"
+)
+
+type certWatcher struct {
+	cert              *tls.Certificate
+	certFile, keyFile string
+	log               *zap.SugaredLogger
+
+	mutex sync.RWMutex
+}
+
+// GetCertificate returns the certificate.
+func (w *certWatcher) GetCertificate(_ *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	w.mutex.RLock()
+	defer w.mutex.RUnlock()
+	return w.cert, nil
+}
+
+// New returns a new cert watcher.
+func New(log *zap.SugaredLogger, certFile, keyFile string) (*certWatcher, error) {
+	w := &certWatcher{
+		certFile: certFile,
+		keyFile:  keyFile,
+		log:      log,
+	}
+	if err := w.loadCertificate(); err != nil {
+		return nil, err
+	}
+	return w, nil
+}
+
+func (w *certWatcher) loadCertificate() error {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	cert, err := tls.LoadX509KeyPair(w.certFile, w.keyFile)
+	if err != nil {
+		return err
+	}
```

Comment:
it seems `w.mutex.Lock()` should be only there, no need to lock it before loading certs

Reply:
> should be only there

should be where?

> no need to lock it before loading certs

Why not? The point is to ensure safety when the secret is changed, and someone tries to read at the same time. By removing `Lock()` from here, you're removing this safety and now readers can potentially read an outdated cert. IMO we need it here.